File: backend/app/models/task.py
from datetime import datetime, timezone

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.database.database import Base


class TaskList(Base):
    __tablename__ = "task_lists"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    name: Mapped[str] = mapped_column(String(80), nullable=False)
    color: Mapped[str] = mapped_column(String(20), default="#4a6ded", nullable=False)
    icon: Mapped[str] = mapped_column(String(40), default="list", nullable=False)
    is_default: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))


class TaskCategory(Base):
    __tablename__ = "task_categories"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    name: Mapped[str] = mapped_column(String(80), nullable=False)
    color: Mapped[str] = mapped_column(String(20), default="#762bbc", nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))


class Task(Base):
    __tablename__ = "tasks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    list_id: Mapped[int | None] = mapped_column(ForeignKey("task_lists.id", ondelete="SET NULL"), nullable=True, index=True)
    category_id: Mapped[int | None] = mapped_column(ForeignKey("task_categories.id", ondelete="SET NULL"), nullable=True, index=True)

    title: Mapped[str] = mapped_column(String(180), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(30), default="not_started", index=True)
    eisenhower: Mapped[str] = mapped_column(String(40), default="important_not_urgent", index=True)
    energy_level: Mapped[str] = mapped_column(String(20), default="medium")

    start_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    due_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, index=True)
    is_all_day: Mapped[bool] = mapped_column(Boolean, default=False)

    repeat_rule: Mapped[str] = mapped_column(String(30), default="none")
    repeat_interval: Mapped[int] = mapped_column(Integer, default=1)
    repeat_until: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    reminder_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    reminder_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, index=True)
    reminder_sent: Mapped[bool] = mapped_column(Boolean, default=False)

    tags: Mapped[list[str]] = mapped_column(JSON, default=list)
    subtasks: Mapped[list[dict]] = mapped_column(JSON, default=list)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

File: frontend/types/task.ts
export type TaskStatus = "not_started" | "in_progress" | "waiting" | "completed";
export type EisenhowerPriority =
  | "urgent_important"
  | "important_not_urgent"
  | "urgent_not_important"
  | "not_urgent_not_important";
export type EnergyLevel = "low" | "medium" | "high";
export type RepeatRule = "none" | "daily" | "weekdays" | "weekly" | "monthly" | "yearly";

export type Subtask = { id: string; title: string; completed: boolean };

export type Task = {
  id: number;
  user_id: number;
  title: string;
  description: string | null;
  list_id: number | null;
  category_id: number | null;
  status: TaskStatus;
  eisenhower: EisenhowerPriority;
  energy_level: EnergyLevel;
  start_at: string | null;
  due_at: string | null;
  is_all_day: boolean;
  repeat_rule: RepeatRule;
  repeat_interval: number;
  repeat_until: string | null;
  reminder_enabled: boolean;
  reminder_at: string | null;
  reminder_sent: boolean;
  tags: string[];
  subtasks: Subtask[];
  completed_at: string | null;
  created_at: string;
  updated_at: string;
};

export type TaskList = {
  id: number;
  user_id: number;
  name: string;
  color: string;
  icon: string;
  is_default: boolean;
  created_at: string;
};

export type TaskCategory = {
  id: number;
  user_id: number;
  name: string;
  color: string;
  created_at: string;
};

export type TaskWorkspace = {
  tasks: Task[];
  lists: TaskList[];
  categories: TaskCategory[];
};

export type TaskPayload = Omit<
  Task,
  "id" | "user_id" | "reminder_sent" | "completed_at" | "created_at" | "updated_at"
>;
